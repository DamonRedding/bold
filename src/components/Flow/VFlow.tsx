import React from 'react'

import { Flow, FlowProps } from './Flow'

export interface VFlowProps extends FlowProps {}

export function VFlow(props: VFlowProps) {
  return <Flow direction='vertical' vSpacing={1} hSpacing={0} {...props} />
}
