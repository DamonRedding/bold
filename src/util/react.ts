import { MutableRefObject, Ref } from 'react'

export function composeRefs<T>(...refs: Array<Ref<T>>) {
    return (value: T) => refs.forEach(ref => setRef(ref, value))
}

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}

/**
 * Returns the next sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getNextSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.nextElementSibling) {
        current = current.nextElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}

/**
 * Returns the previous sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getPreviousSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.previousElementSibling) {
        current = current.previousElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}
