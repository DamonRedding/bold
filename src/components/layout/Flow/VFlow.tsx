import React from 'react'

import { withStyles } from '../../../styles'

import { Flow, FlowProps } from './Flow'

export interface VFlowProps extends FlowProps {
}

@withStyles
export class VFlow extends React.PureComponent<VFlowProps> {

    render() {
        return (
            <Flow direction='vertical' vSpacing={1} hSpacing={0} {...this.props}>
                {this.props.children}
            </Flow>
        )
    }
}
