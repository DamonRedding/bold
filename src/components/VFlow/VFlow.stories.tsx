import { number, select } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import React from 'react'

import { Button } from '../Button'
import { AlignItems, JustifyContent } from '../Grid/Grid'

import { VFlow } from './VFlow'

const justifyContentOptions: JustifyContent[] = [
  'flex-start',
  'flex-end',
  'center',
  'space-between',
  'space-around',
  'space-evenly',
]

const alignItemsOptions: AlignItems[] = ['flex-start', 'flex-end', 'center', 'baseline', 'stretch']

storiesOf('Components|VFlow', module).add('default', () => (
  <VFlow
    vSpacing={number('vSpacing', 1)}
    alignItems={select('alignItems', alignItemsOptions, 'stretch')}
    justifyContent={select('justifyContent', justifyContentOptions, 'flex-start')}
  >
    <Button kind='primary'>Button 1</Button>
    <Button>Button 2</Button>
    <Button>Button 3</Button>
  </VFlow>
))
