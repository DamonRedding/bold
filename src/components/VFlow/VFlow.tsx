import React, { CSSProperties } from 'react'

import { ExternalStyles, Theme, useStyles } from '../../styles'
import { Omit } from '../../util'
import { AlignItems, JustifyContent } from '../Grid'

export interface VFlowProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'style'> {
  vSpacing?: number
  alignItems?: AlignItems
  justifyContent?: JustifyContent
  style?: ExternalStyles
}

export function VFlow(props: VFlowProps) {
  const { vSpacing, alignItems, justifyContent, style, ...rest } = props
  const { classes, css } = useStyles(createStyles, props)

  return <div className={css(classes.container, style)} {...rest} />
}

VFlow.defaultProps = {
  vSpacing: 1,
} as VFlowProps

const createStyles = (theme: Theme, { vSpacing, alignItems, justifyContent }: VFlowProps) => ({
  container: {
    display: 'grid',
    gridAutoFlow: 'row',
    gridGap: vSpacing ? `${vSpacing}rem` : undefined,
    alignItems,
    justifyContent,
  } as CSSProperties,
})
