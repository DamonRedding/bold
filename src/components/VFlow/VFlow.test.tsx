import { render } from '@testing-library/react'
import React from 'react'

import { VFlow } from './VFlow'

it('should render correctly', () => {
  const { container } = render(
    <VFlow>
      <span>1</span>
      <span>2</span>
      <span>3</span>
      456
    </VFlow>
  )
  expect(container).toMatchSnapshot()
})

it('should accept some grid props', () => {
  const { container } = render(
    <VFlow vSpacing={2} alignItems='center' justifyContent='space-around'>
      <span>1</span>
      <span>2</span>
      <span>3</span>
    </VFlow>
  )
  expect(container).toMatchSnapshot()
})
