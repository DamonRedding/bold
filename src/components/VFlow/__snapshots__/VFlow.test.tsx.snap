// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept some grid props 1`] = `
.emotion-0 {
  display: grid;
  grid-auto-flow: row;
  grid-gap: 2rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
}

<div>
  <div
    class="emotion-0"
  >
    <span>
      1
    </span>
    <span>
      2
    </span>
    <span>
      3
    </span>
  </div>
</div>
`;

exports[`should render correctly 1`] = `
.emotion-0 {
  display: grid;
  grid-auto-flow: row;
  grid-gap: 1rem;
}

<div>
  <div
    class="emotion-0"
  >
    <span>
      1
    </span>
    <span>
      2
    </span>
    <span>
      3
    </span>
    456
  </div>
</div>
`;
